# **1. Introduction**


**Project Name : Manage Finance Division Data Using PostgreSQL Server**

**Author       : Tasya Amalia**






# **2. Import Libraries**

In [1]:
# import pandas
import pandas as pd

# **3. Data Loading**

In [2]:
'''
Melakukan load  data ke dalam pandas DataFrame dari sumber csv
'''
df = pd.read_csv('https://raw.githubusercontent.com/FTDS-learning-materials/phase-0/main/src/Financials.csv')

# Menampilkan data teratas
df.head()


Segment  Country     Product   Discount Band   Units Sold   \
0  Government   Canada   Carretera            None    $1,618.50    
1  Government  Germany   Carretera            None    $1,321.00    
2   Midmarket   France   Carretera            None    $2,178.00    
3   Midmarket  Germany   Carretera            None      $888.00    
4   Midmarket   Mexico   Carretera            None    $2,470.00    

   Manufacturing Price   Sale Price   Gross Sales   Discounts         Sales   \
0                $3.00       $20.00    $32,370.00        $-      $32,370.00    
1                $3.00       $20.00    $26,420.00        $-      $26,420.00    
2                $3.00       $15.00    $32,670.00        $-      $32,670.00    
3                $3.00       $15.00    $13,320.00        $-      $13,320.00    
4                $3.00       $15.00    $37,050.00        $-      $37,050.00    

          COGS        Profit         Date  Month Number  Month Name   Year  
0   $16,185.00    $16,185.00   01/01/2014             1     January   2014  
1   $13,210.00    $13,210.00   01/01/2014             1     January   2014  
2   $21,780.00    $10,890.00   01/06/2014             6        June   2014  
3    $8,880.00     $4,440.00   01/06/2014             6        June   2014  
4   $24,700.00    $12,350.00   01/06/2014             6        June   2014

- Pada tabel diatas menunjukkan informasi mengenai pendapatan dan profit per produk, segmentasi, dan negara yang didapatkan tiap harinya. Dari tabel tersebut diketahui data ini belum bersih dan siap untuk dimasukkan ke dalam database untuk di analisis lebih lanjut. Langkah selanjutnya yaitu mengexplore data untuk mengetahui isi data nya dan melakukan cleaning data.

- Hasil pada tabel tersebut menunjukkan terdapat simbol mata uang ($) di beberapa kolom dan karakter lainnya seperti '-' dan '()'. Oleh karena itu diperlukan data cleaning untuk memudahkan proses penggantian tipe data yang sesuai.


# **4. Data Exploration**

In [3]:
# menampilkan summary informasi dataframe
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 700 entries, 0 to 699
Data columns (total 16 columns):
 #   Column                 Non-Null Count  Dtype 
---  ------                 --------------  ----- 
 0   Segment                700 non-null    object
 1   Country                700 non-null    object
 2    Product               700 non-null    object
 3    Discount Band         700 non-null    object
 4    Units Sold            700 non-null    object
 5    Manufacturing Price   700 non-null    object
 6    Sale Price            700 non-null    object
 7    Gross Sales           700 non-null    object
 8    Discounts             700 non-null    object
 9     Sales                700 non-null    object
 10   COGS                  700 non-null    object
 11   Profit                700 non-null    object
 12  Date                   700 non-null    object
 13  Month Number           700 non-null    int64 
 14   Month Name            700 non-null    object
 15  Year                   

- Ringkasan informasi dari DataFrame ini yaitu terdapat 700 baris dan 16 kolom. DataFrame ini memiliki beberapa tabel yaitu, segment, country, product, discount band, units sold, manufacturing price, sale price, gross sales, discounts, sales, COGS, profit, date, month number, month name, dan year.

- Hasil diatas menunjukkan bahwa tidak ada nilai null di setiap kolom.

- Tipe data dari DataFrame ini terdiri dari 14 object dan 2 integer. Namun perlu diperhatikan lebih lanjut, terdapat kesalahan tipe data seperti units sold, manufacturing price, sale price, gross sales, discounts, sales, COGS, profit yang seharusnya float dan date yang seharusnya date time.

- Oleh karena itu, diperlukan data cleaning untuk proses selanjutnya.

In [4]:
# menampilkan kolom yang terdapat pada DataFrame
df.columns

Index(['Segment', 'Country', ' Product ', ' Discount Band ', ' Units Sold ',
       ' Manufacturing Price ', ' Sale Price ', ' Gross Sales ', ' Discounts ',
       '  Sales ', ' COGS ', ' Profit ', 'Date', 'Month Number',
       ' Month Name ', 'Year'],
      dtype='object')

- Pada hasil tersebut menunjukkan bahwa terdapat 16 kolom, yaitu segment, country, product, discount band, units sold, manufacturing price, sale price, gross sales, discounts, sales, COGS, profit, date, month number, month name, dan year.

- Penulisan nama kolom tersebut masih memiliki whitespace yang harus dihilangkan dengan proses data cleaning.


In [5]:
# menampilkan nilai unik yang ada di kolom Segment
df['Segment'].unique()

array(['Government', 'Midmarket', 'Channel Partners', 'Enterprise',
       'Small Business'], dtype=object)

Pada hasil tersebut diketahui kolom 'segment' memiliki nilai unik, yaitu Government, Midmarket, Channel Partners, Enterprise, dan Small Business.

In [6]:
# menampilkan nilai unik yang ada di kolom Country
df['Country'].unique()

array(['Canada', 'Germany', 'France', 'Mexico',
       'United States of America'], dtype=object)

Pada hasil tersebut diketahui kolom 'country' memiliki nilai unik, yaitu Canada, Germany, France, Mexico, dan United States of America.

In [7]:
# menampilkan nilai unik yang ada di kolom Product
df[' Product '].unique()

array([' Carretera ', ' Montana ', ' Paseo ', ' Velo ', ' VTT ',
       ' Amarilla '], dtype=object)

- Pada hasil tersebut diketahui kolom 'product' memiliki nilai unik, yaitu carretera, montana, paseo, velo, VTT, dan amarilla.

- Penulisan nilai unik di kolom product terdapat whitespace sehingga perlu dihapus dengan proses data cleaning.

In [8]:
# menampilkan nilai unik yang ada di kolom Discount Band
df[' Discount Band '].unique()

array([' None ', ' Low ', ' Medium ', ' High '], dtype=object)

- Pada hasil tersebut diketahui kolom 'discount band' memiliki nilai unik, yaitu none, low, medium, dan high.

- Penulisan nilai unik di kolom discount band terdapat whitespace sehingga perlu dihapus dengan proses data cleaning.

# **5. Data Cleaning**

##### a. Mengganti nama kolom

In [9]:
'''
Mengubah penulisan nama kolom menjadi huruf kecil (lower)
'''

df.rename(str.lower, axis='columns', inplace=True)
df.head()

segment  country     product   discount band   units sold   \
0  Government   Canada   Carretera            None    $1,618.50    
1  Government  Germany   Carretera            None    $1,321.00    
2   Midmarket   France   Carretera            None    $2,178.00    
3   Midmarket  Germany   Carretera            None      $888.00    
4   Midmarket   Mexico   Carretera            None    $2,470.00    

   manufacturing price   sale price   gross sales   discounts         sales   \
0                $3.00       $20.00    $32,370.00        $-      $32,370.00    
1                $3.00       $20.00    $26,420.00        $-      $26,420.00    
2                $3.00       $15.00    $32,670.00        $-      $32,670.00    
3                $3.00       $15.00    $13,320.00        $-      $13,320.00    
4                $3.00       $15.00    $37,050.00        $-      $37,050.00    

          cogs        profit         date  month number  month name   year  
0   $16,185.00    $16,185.00   01/01/2014             1     January   2014  
1   $13,210.00    $13,210.00   01/01/2014             1     January   2014  
2   $21,780.00    $10,890.00   01/06/2014             6        June   2014  
3    $8,880.00     $4,440.00   01/06/2014             6        June   2014  
4   $24,700.00    $12,350.00   01/06/2014             6        June   2014

 Langkah pertama untuk proses data cleaning yaitu manipulasi kolom dengan penggantian nama kolom menjadi huruf kecil.

##### b. Menghapus Whitespace pada nama kolom

In [10]:
# Menghapus whitespace yang terdapat di nama kolom 
df.columns = df.columns.str.strip()
df.columns

Index(['segment', 'country', 'product', 'discount band', 'units sold',
       'manufacturing price', 'sale price', 'gross sales', 'discounts',
       'sales', 'cogs', 'profit', 'date', 'month number', 'month name',
       'year'],
      dtype='object')

Penghapusan whitespace yang terdapat di nama kolom untuk menghindari error saat analisis data.

##### c. Menghapus Whitespace pada value kolom (nilai unik kolom)

In [11]:
# menghapus semua whitespace value pada kolom dengan tipe data object

for col in df.columns:
    try:
        df[col] = df[col].str.strip()
    except AttributeError:
        pass


##### d. Menghapus karakter '-' pada  kolom discounts

In [16]:
# Menghapus karakter '-' di kolom discounts
df['discounts'] = df['discounts'].str.replace('-', '0')
df

segment                   country    product discount band  \
0          Government                    Canada  Carretera          None   
1          Government                   Germany  Carretera          None   
2           Midmarket                    France  Carretera          None   
3           Midmarket                   Germany  Carretera          None   
4           Midmarket                    Mexico  Carretera          None   
..                ...                       ...        ...           ...   
695    Small Business                    France   Amarilla          High   
696    Small Business                    Mexico   Amarilla          High   
697        Government                    Mexico    Montana          High   
698        Government                    Canada      Paseo          High   
699  Channel Partners  United States of America        VTT          High   

    units sold manufacturing price sale price   gross sales     discounts  \
0    $1,618.50               $3.00     $20.00    $32,370.00            $0   
1    $1,321.00               $3.00     $20.00    $26,420.00            $0   
2    $2,178.00               $3.00     $15.00    $32,670.00            $0   
3      $888.00               $3.00     $15.00    $13,320.00            $0   
4    $2,470.00               $3.00     $15.00    $37,050.00            $0   
..         ...                 ...        ...           ...           ...   
695  $2,475.00             $260.00    $300.00  $7,42,500.00  $1,11,375.00   
696    $546.00             $260.00    $300.00  $1,63,800.00    $24,570.00   
697  $1,368.00               $5.00      $7.00     $9,576.00     $1,436.40   
698    $723.00              $10.00      $7.00     $5,061.00       $759.15   
699  $1,806.00             $250.00     $12.00    $21,672.00     $3,250.80   

            sales          cogs      profit        date  month number  \
0      $32,370.00    $16,185.00  $16,185.00  01/01/2014             1   
1      $26,420.00    $13,210.00  $13,210.00  01/01/2014             1   
2      $32,670.00    $21,780.00  $10,890.00  01/06/2014             6   
3      $13,320.00     $8,880.00   $4,440.00  01/06/2014             6   
4      $37,050.00    $24,700.00  $12,350.00  01/06/2014             6   
..            ...           ...         ...         ...           ...   
695  $6,31,125.00  $6,18,750.00  $12,375.00  01/03/2014             3   
696  $1,39,230.00  $1,36,500.00   $2,730.00  01/10/2014            10   
697     $8,139.60     $6,840.00   $1,299.60  01/02/2014             2   
698     $4,301.85     $3,615.00     $686.85  01/04/2014             4   
699    $18,421.20     $5,418.00  $13,003.20  01/05/2014             5   

    month name  year  
0      January  2014  
1      January  2014  
2         June  2014  
3         June  2014  
4         June  2014  
..         ...   ...  
695      March  2014  
696    October  2014  
697   February  2014  
698      April  2014  
699        May  2014  

[700 rows x 16 columns]

In [17]:
# Menampilkan value kolom discounts setelah data cleaning
df['discounts'].unique()

array(['$0', '$276.15', '$344.40', '$72.10', '$44.73', '$92.82',
       '$222.96', '$4,235.00', '$177.03', '$173.40', '$412.50', '$320.52',
       '$91.92', '$1,482.00', '$4,889.50', '$7,542.50', '$332.10',
       '$6,903.00', '$275.10', '$128.10', '$7,494.00', '$828.75',
       '$227.10', '$314.48', '$908.75', '$983.75', '$2,278.75', '$112.05',
       '$8,715.00', '$772.80', '$25.34', '$1,153.75', '$146.44', '$18.41',
       '$3,302.25', '$2,958.00', '$2,180.00', '$238.68', '$48.15',
       '$1,856.25', '$310.80', '$1,284.00', '$300.30', '$19,964.00',
       '$274.08', '$626.40', '$165.60', '$4,150.00', '$708.90',
       '$5,508.00', '$10,368.00', '$1,655.00', '$2,022.50', '$5,362.50',
       '$428.40', '$11,496.00', '$6,822.50', '$577.50', '$281.82',
       '$253.20', '$260.16', '$20,762.00', '$20,139.00', '$217.60',
       '$1,862.00', '$13,580.00', '$1,554.00', '$6,606.00', '$5,690.00',
       '$494.40', '$376.40', '$15,913.13', '$1,548.00', '$6,201.00',
       '$700.92', '$326.88'

Hasil data cleaning pada kolom discounts menunjukkan bahwa karakter '-' sudah diganti dengan nilai 0

##### e. Menghilangkan karakter '-' pada value kolom profit

In [18]:
# Menghapus karakter '-' di kolom profit
df['profit'] = df['profit'].str.replace('-', '0')
df

segment                   country    product discount band  \
0          Government                    Canada  Carretera          None   
1          Government                   Germany  Carretera          None   
2           Midmarket                    France  Carretera          None   
3           Midmarket                   Germany  Carretera          None   
4           Midmarket                    Mexico  Carretera          None   
..                ...                       ...        ...           ...   
695    Small Business                    France   Amarilla          High   
696    Small Business                    Mexico   Amarilla          High   
697        Government                    Mexico    Montana          High   
698        Government                    Canada      Paseo          High   
699  Channel Partners  United States of America        VTT          High   

    units sold manufacturing price sale price   gross sales     discounts  \
0    $1,618.50               $3.00     $20.00    $32,370.00            $0   
1    $1,321.00               $3.00     $20.00    $26,420.00            $0   
2    $2,178.00               $3.00     $15.00    $32,670.00            $0   
3      $888.00               $3.00     $15.00    $13,320.00            $0   
4    $2,470.00               $3.00     $15.00    $37,050.00            $0   
..         ...                 ...        ...           ...           ...   
695  $2,475.00             $260.00    $300.00  $7,42,500.00  $1,11,375.00   
696    $546.00             $260.00    $300.00  $1,63,800.00    $24,570.00   
697  $1,368.00               $5.00      $7.00     $9,576.00     $1,436.40   
698    $723.00              $10.00      $7.00     $5,061.00       $759.15   
699  $1,806.00             $250.00     $12.00    $21,672.00     $3,250.80   

            sales          cogs      profit        date  month number  \
0      $32,370.00    $16,185.00  $16,185.00  01/01/2014             1   
1      $26,420.00    $13,210.00  $13,210.00  01/01/2014             1   
2      $32,670.00    $21,780.00  $10,890.00  01/06/2014             6   
3      $13,320.00     $8,880.00   $4,440.00  01/06/2014             6   
4      $37,050.00    $24,700.00  $12,350.00  01/06/2014             6   
..            ...           ...         ...         ...           ...   
695  $6,31,125.00  $6,18,750.00  $12,375.00  01/03/2014             3   
696  $1,39,230.00  $1,36,500.00   $2,730.00  01/10/2014            10   
697     $8,139.60     $6,840.00   $1,299.60  01/02/2014             2   
698     $4,301.85     $3,615.00     $686.85  01/04/2014             4   
699    $18,421.20     $5,418.00  $13,003.20  01/05/2014             5   

    month name  year  
0      January  2014  
1      January  2014  
2         June  2014  
3         June  2014  
4         June  2014  
..         ...   ...  
695      March  2014  
696    October  2014  
697   February  2014  
698      April  2014  
699        May  2014  

[700 rows x 16 columns]

In [19]:
# Menampilkan value kolom profit setelah data cleaning
df['profit'].unique()

array(['$16,185.00', '$13,210.00', '$10,890.00', '$4,440.00',
       '$12,350.00', '$1,36,170.00', '$4,605.00', '$22,662.00',
       '$18,990.00', '$13,905.00', '$13,327.50', '$47,900.00',
       '$4,292.00', '$1,725.00', '$3,075.00', '$2,920.00', '$4,870.00',
       '$90,540.00', '$3,303.00', '$1,766.00', '$2,745.00', '$39,400.00',
       '$12,360.00', '$2,286.00', '$1,55,250.00', '$8,208.00',
       '$10,760.00', '$18,170.00', '$2,986.00', '$9,020.00', '$19,449.00',
       '$14,105.00', '$1,00,050.00', '$25,542.00', '$1,37,430.00',
       '$1,07,550.00', '$2,47,500.00', '$17,577.00', '$21,097.50',
       '$3,372.00', '$19,269.00', '$7,613.85', '$11,135.60', '$1,987.90',
       '$1,233.27', '$2,559.18', '$16,499.04', '$1,04,665.00',
       '$4,880.97', '$12,831.60', '$1,237.50', '$23,718.48', '$6,802.08',
       '$23,218.00', '$1,20,840.50', '$1,86,407.50', '$10,737.90',
       '$1,08,147.00', '$13,479.90', '$3,531.90', '$1,17,406.00',
       '$2,486.25', '$7,342.90', '$8,670.53', '$2

Hasil data cleaning pada kolom profit menunjukkan bahwa karakter '-' sudah diganti dengan nilai 0. Namun masih terdapat karakter () yang selanjutnya akan dilakukan data cleaning lebih lanjut.

##### f. Menghilangkan karakter ')' dan menggantikan karakter '(' menjadi '-' pada value kolom profit

In [20]:
# Menghapus karakter ')' dan menggantikan karakter '(' menjadi '-' pada value kolom profit
df['profit'] = df['profit'].str.replace('(', '-').str.replace(')', '')


print(df)

              segment                   country    product discount band  \
0          Government                    Canada  Carretera          None   
1          Government                   Germany  Carretera          None   
2           Midmarket                    France  Carretera          None   
3           Midmarket                   Germany  Carretera          None   
4           Midmarket                    Mexico  Carretera          None   
..                ...                       ...        ...           ...   
695    Small Business                    France   Amarilla          High   
696    Small Business                    Mexico   Amarilla          High   
697        Government                    Mexico    Montana          High   
698        Government                    Canada      Paseo          High   
699  Channel Partners  United States of America        VTT          High   

    units sold manufacturing price sale price   gross sales     discounts  \
0    $1,61

C:\Users\irfan\AppData\Local\Temp\ipykernel_24616\775344527.py:2: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  df['profit'] = df['profit'].str.replace('(', '-').str.replace(')', '')


In [21]:
# Menampilkan value kolom profit setelah data cleaning
df['profit'].unique()

array(['$16,185.00', '$13,210.00', '$10,890.00', '$4,440.00',
       '$12,350.00', '$1,36,170.00', '$4,605.00', '$22,662.00',
       '$18,990.00', '$13,905.00', '$13,327.50', '$47,900.00',
       '$4,292.00', '$1,725.00', '$3,075.00', '$2,920.00', '$4,870.00',
       '$90,540.00', '$3,303.00', '$1,766.00', '$2,745.00', '$39,400.00',
       '$12,360.00', '$2,286.00', '$1,55,250.00', '$8,208.00',
       '$10,760.00', '$18,170.00', '$2,986.00', '$9,020.00', '$19,449.00',
       '$14,105.00', '$1,00,050.00', '$25,542.00', '$1,37,430.00',
       '$1,07,550.00', '$2,47,500.00', '$17,577.00', '$21,097.50',
       '$3,372.00', '$19,269.00', '$7,613.85', '$11,135.60', '$1,987.90',
       '$1,233.27', '$2,559.18', '$16,499.04', '$1,04,665.00',
       '$4,880.97', '$12,831.60', '$1,237.50', '$23,718.48', '$6,802.08',
       '$23,218.00', '$1,20,840.50', '$1,86,407.50', '$10,737.90',
       '$1,08,147.00', '$13,479.90', '$3,531.90', '$1,17,406.00',
       '$2,486.25', '$7,342.90', '$8,670.53', '$2

Hasil data cleaning pada kolom profit menunjukkan bahwa karakter ')' sudah di hilangkan dan karakter '(' diganti minus (-) karena karakter tanda kurung () di data keuangan menunjukkan nilai minus atau rugi.

##### g. Menghilangkan karakter '$' pada beberapa kolom 

In [22]:
'''
Melakukan looping untuk memeriksa apakah ada karakter '$' pada kolom dengan tipe data object
'''

for col in df.columns:
    if df[col].dtype == 'object':  # Memeriksa apakah kolom merupakan tipe data objek (string)
        df[col] = df[col].str.replace('$', '')

print(df)

              segment                   country    product discount band  \
0          Government                    Canada  Carretera          None   
1          Government                   Germany  Carretera          None   
2           Midmarket                    France  Carretera          None   
3           Midmarket                   Germany  Carretera          None   
4           Midmarket                    Mexico  Carretera          None   
..                ...                       ...        ...           ...   
695    Small Business                    France   Amarilla          High   
696    Small Business                    Mexico   Amarilla          High   
697        Government                    Mexico    Montana          High   
698        Government                    Canada      Paseo          High   
699  Channel Partners  United States of America        VTT          High   

    units sold manufacturing price sale price  gross sales    discounts  \
0     1,618.

C:\Users\irfan\AppData\Local\Temp\ipykernel_24616\2878011141.py:7: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  df[col] = df[col].str.replace('$', '')


In [23]:
# Menampilkan hasil dari penghapusan karakter '$' pada kolom dengan tipe data object
df.head()

segment  country    product discount band units sold  \
0  Government   Canada  Carretera          None   1,618.50   
1  Government  Germany  Carretera          None   1,321.00   
2   Midmarket   France  Carretera          None   2,178.00   
3   Midmarket  Germany  Carretera          None     888.00   
4   Midmarket   Mexico  Carretera          None   2,470.00   

  manufacturing price sale price gross sales discounts      sales       cogs  \
0                3.00      20.00   32,370.00         0  32,370.00  16,185.00   
1                3.00      20.00   26,420.00         0  26,420.00  13,210.00   
2                3.00      15.00   32,670.00         0  32,670.00  21,780.00   
3                3.00      15.00   13,320.00         0  13,320.00   8,880.00   
4                3.00      15.00   37,050.00         0  37,050.00  24,700.00   

      profit        date  month number month name  year  
0  16,185.00  01/01/2014             1    January  2014  
1  13,210.00  01/01/2014             1    January  2014  
2  10,890.00  01/06/2014             6       June  2014  
3   4,440.00  01/06/2014             6       June  2014  
4  12,350.00  01/06/2014             6       June  2014

##### h. Menghapus karakter (,) pada beberapa kolom 

In [24]:
# List kolom yang akan diubah
columns_to_clean = ['units sold','gross sales','discounts','sales','cogs','profit']

# Menghapus karakter koma (,) dari isi kolom yang diinginkan
for col in columns_to_clean:
    df[col] = df[col].str.replace(',', '')

print(df)

# Menampilkan hasil dari penghapusan karakter ',' pada beberapa kolom
df.head()

              segment                   country    product discount band  \
0          Government                    Canada  Carretera          None   
1          Government                   Germany  Carretera          None   
2           Midmarket                    France  Carretera          None   
3           Midmarket                   Germany  Carretera          None   
4           Midmarket                    Mexico  Carretera          None   
..                ...                       ...        ...           ...   
695    Small Business                    France   Amarilla          High   
696    Small Business                    Mexico   Amarilla          High   
697        Government                    Mexico    Montana          High   
698        Government                    Canada      Paseo          High   
699  Channel Partners  United States of America        VTT          High   

    units sold manufacturing price sale price gross sales  discounts  \
0      1618.50 

segment  country    product discount band units sold  \
0  Government   Canada  Carretera          None    1618.50   
1  Government  Germany  Carretera          None    1321.00   
2   Midmarket   France  Carretera          None    2178.00   
3   Midmarket  Germany  Carretera          None     888.00   
4   Midmarket   Mexico  Carretera          None    2470.00   

  manufacturing price sale price gross sales discounts     sales      cogs  \
0                3.00      20.00    32370.00         0  32370.00  16185.00   
1                3.00      20.00    26420.00         0  26420.00  13210.00   
2                3.00      15.00    32670.00         0  32670.00  21780.00   
3                3.00      15.00    13320.00         0  13320.00   8880.00   
4                3.00      15.00    37050.00         0  37050.00  24700.00   

     profit        date  month number month name  year  
0  16185.00  01/01/2014             1    January  2014  
1  13210.00  01/01/2014             1    January  2014  
2  10890.00  01/06/2014             6       June  2014  
3   4440.00  01/06/2014             6       June  2014  
4  12350.00  01/06/2014             6       June  2014

Penghapusan karakter (,) untuk menghindarkan error saat mengubah tipe data yang seharusnya

#### i. Mengubah tipe data yang seharusnya

In [25]:
# List tipe data kolom yang akan diubah
columns_to_convert = ['units sold','manufacturing price','sale price','gross sales','discounts','sales','cogs','profit']

# Mengubah tipe data kolom dari object ke float
df[columns_to_convert] = df[columns_to_convert].astype(float)


# Mengubah tipe data kolom 'date' menjadi datetime
df['date'] = pd.to_datetime(df['date'])

In [26]:
# Menampilkan perubaha tipe data
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 700 entries, 0 to 699
Data columns (total 16 columns):
 #   Column               Non-Null Count  Dtype         
---  ------               --------------  -----         
 0   segment              700 non-null    object        
 1   country              700 non-null    object        
 2   product              700 non-null    object        
 3   discount band        700 non-null    object        
 4   units sold           700 non-null    float64       
 5   manufacturing price  700 non-null    float64       
 6   sale price           700 non-null    float64       
 7   gross sales          700 non-null    float64       
 8   discounts            700 non-null    float64       
 9   sales                700 non-null    float64       
 10  cogs                 700 non-null    float64       
 11  profit               700 non-null    float64       
 12  date                 700 non-null    datetime64[ns]
 13  month number         700 non-null  

Ringkasan informasi menunjukkan bahwa tipe data sudah diubah sesuai yang seharusnya, yaitu terdapat tipe data object sebanyak 5 kolom, float sebanyak 8 kolom, datetime sebanyak 1 kolom dan integer sebanyak 2 kolom.

#### j. Memeriksa missing value

In [27]:
# Memeriksa missing value
missing_values = df.isna().sum()
print("Jumlah missing value pada setiap kolom:")
print(missing_values)

# Handling missing value dengan imputasi nilai 0
df.fillna(0, inplace=True)

print("\nDataFrame setelah handling missing value:")
print(df)


Jumlah missing value pada setiap kolom:
segment                0
country                0
product                0
discount band          0
units sold             0
manufacturing price    0
sale price             0
gross sales            0
discounts              0
sales                  0
cogs                   0
profit                 0
date                   0
month number           0
month name             0
year                   0
dtype: int64

DataFrame setelah handling missing value:
              segment                   country    product discount band  \
0          Government                    Canada  Carretera          None   
1          Government                   Germany  Carretera          None   
2           Midmarket                    France  Carretera          None   
3           Midmarket                   Germany  Carretera          None   
4           Midmarket                    Mexico  Carretera          None   
..                ...                       ... 

Hasil data cleaning menunjukkan data ini tidak mimiliki missing value sehingga dapat dilanjutkan proses analisis data.

#### k. Menghapus beberapa kolom yang tidak diperlukan

In [28]:
# menghapus kolom 
df.drop(['month number', 'month name', 'year'], axis=1, inplace=True)

In [29]:
# Menampilkan hasil dari penghapusan kolom month number, month name, dan year
df.head()

segment  country    product discount band  units sold  \
0  Government   Canada  Carretera          None      1618.5   
1  Government  Germany  Carretera          None      1321.0   
2   Midmarket   France  Carretera          None      2178.0   
3   Midmarket  Germany  Carretera          None       888.0   
4   Midmarket   Mexico  Carretera          None      2470.0   

   manufacturing price  sale price  gross sales  discounts    sales     cogs  \
0                  3.0        20.0      32370.0        0.0  32370.0  16185.0   
1                  3.0        20.0      26420.0        0.0  26420.0  13210.0   
2                  3.0        15.0      32670.0        0.0  32670.0  21780.0   
3                  3.0        15.0      13320.0        0.0  13320.0   8880.0   
4                  3.0        15.0      37050.0        0.0  37050.0  24700.0   

    profit       date  
0  16185.0 2014-01-01  
1  13210.0 2014-01-01  
2  10890.0 2014-01-06  
3   4440.0 2014-01-06  
4  12350.0 2014-01-06

In [30]:
# Menampilkan ringkasan informasi setelah data cleaning
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 700 entries, 0 to 699
Data columns (total 13 columns):
 #   Column               Non-Null Count  Dtype         
---  ------               --------------  -----         
 0   segment              700 non-null    object        
 1   country              700 non-null    object        
 2   product              700 non-null    object        
 3   discount band        700 non-null    object        
 4   units sold           700 non-null    float64       
 5   manufacturing price  700 non-null    float64       
 6   sale price           700 non-null    float64       
 7   gross sales          700 non-null    float64       
 8   discounts            700 non-null    float64       
 9   sales                700 non-null    float64       
 10  cogs                 700 non-null    float64       
 11  profit               700 non-null    float64       
 12  date                 700 non-null    datetime64[ns]
dtypes: datetime64[ns](1), float64(8), o

Hasil tersebut menunjukkan setelah dilakukan data cleaning, DataFrame ini memiliki 700 baris dan 13 kolom dengan tipe data yang sudah seharusnya. DataFrame ini memiliki tipe data float sebanyak 8 data, object sebanyak 4 data dan datetime sebanyak 1 data.

# **5. Data Normalization**

Pada tahap ini dilakukan pembuatan DataFrame baru dimana terdapat 4 tabel yang  masing-masing tabel memiliki primary key, yaitu tabel segment, county, product, dan discount band. Tabel utama memiliki foreign key dari keempat tabel tersebut.

#### a. Menambahkan kolom baru

In [31]:
# Menampilkan Tabel utama untuk merge
df.head()

segment  country    product discount band  units sold  \
0  Government   Canada  Carretera          None      1618.5   
1  Government  Germany  Carretera          None      1321.0   
2   Midmarket   France  Carretera          None      2178.0   
3   Midmarket  Germany  Carretera          None       888.0   
4   Midmarket   Mexico  Carretera          None      2470.0   

   manufacturing price  sale price  gross sales  discounts    sales     cogs  \
0                  3.0        20.0      32370.0        0.0  32370.0  16185.0   
1                  3.0        20.0      26420.0        0.0  26420.0  13210.0   
2                  3.0        15.0      32670.0        0.0  32670.0  21780.0   
3                  3.0        15.0      13320.0        0.0  13320.0   8880.0   
4                  3.0        15.0      37050.0        0.0  37050.0  24700.0   

    profit       date  
0  16185.0 2014-01-01  
1  13210.0 2014-01-01  
2  10890.0 2014-01-06  
3   4440.0 2014-01-06  
4  12350.0 2014-01-06

### b. Membuat Beberapa Tabel Baru

In [32]:
# Membuat DataFrame Segment
data = {'id segment': [1, 2, 3, 4, 5],
        'segment': ['Government', 'Midmarket', 'Channel Partners', 'Enterprise', 'Small Business']}
df1 = pd.DataFrame(data)
display(df1)


id segment           segment
0           1        Government
1           2         Midmarket
2           3  Channel Partners
3           4        Enterprise
4           5    Small Business

Pada tabel segment memiliki nilai unik dan primary key pada kolom id segment

In [33]:
# Membuat DataFrame Country
data = {'id country': [1, 2, 3, 4, 5],
        'country': ['Canada', 'Germany', 'France', 'Mexico', 'United States of America']}
df2 = pd.DataFrame(data)
display(df2)


id country                   country
0           1                    Canada
1           2                   Germany
2           3                    France
3           4                    Mexico
4           5  United States of America

Pada tabel country memiliki nilai unik dan primary key pada kolom id_country

In [34]:
# Membuat DataFrame Product
data = {'id product': [1, 2, 3, 4, 5, 6],
        'product': ['Carretera', 'Montana', 'Paseo', 'Velo', 'VTT', 'Amarilla']}
df3 = pd.DataFrame(data)
display(df3)


id product    product
0           1  Carretera
1           2    Montana
2           3      Paseo
3           4       Velo
4           5        VTT
5           6   Amarilla

Pada tabel product memiliki nilai unik dan primary key pada kolom id_product

In [35]:
# Membuat DataFrame Discount Band
data = {'id discount band': [1, 2, 3, 4],
        'discount band': ['None', 'Low', 'Medium', 'High']}
df4 = pd.DataFrame(data)
display(df4)


id discount band discount band
0                 1          None
1                 2           Low
2                 3        Medium
3                 4          High

Pada tabel discount band memiliki nilai unik dan primary key pada kolom id_discount band

### c. Menggabungkan 4 tabel yang memiliki primary key ke dalam tabel utama yang memiliki foreign key dari masing-masing kolom(segment, country, product, dan discount band)

In [36]:
'''
Menggabungkan dua DataFrame berdasarkan kolom 'segment' menggunakan left join
'''

merge_df = pd.merge(df,df1, on ='segment', how='left')

In [37]:
# Menampilkan hasil gabungan tabel utama dengan tabel segment 
merge_df

segment                   country    product discount band  \
0          Government                    Canada  Carretera          None   
1          Government                   Germany  Carretera          None   
2           Midmarket                    France  Carretera          None   
3           Midmarket                   Germany  Carretera          None   
4           Midmarket                    Mexico  Carretera          None   
..                ...                       ...        ...           ...   
695    Small Business                    France   Amarilla          High   
696    Small Business                    Mexico   Amarilla          High   
697        Government                    Mexico    Montana          High   
698        Government                    Canada      Paseo          High   
699  Channel Partners  United States of America        VTT          High   

     units sold  manufacturing price  sale price  gross sales  discounts  \
0        1618.5                  3.0        20.0      32370.0       0.00   
1        1321.0                  3.0        20.0      26420.0       0.00   
2        2178.0                  3.0        15.0      32670.0       0.00   
3         888.0                  3.0        15.0      13320.0       0.00   
4        2470.0                  3.0        15.0      37050.0       0.00   
..          ...                  ...         ...          ...        ...   
695      2475.0                260.0       300.0     742500.0  111375.00   
696       546.0                260.0       300.0     163800.0   24570.00   
697      1368.0                  5.0         7.0       9576.0    1436.40   
698       723.0                 10.0         7.0       5061.0     759.15   
699      1806.0                250.0        12.0      21672.0    3250.80   

         sales      cogs    profit       date  id segment  
0     32370.00   16185.0  16185.00 2014-01-01           1  
1     26420.00   13210.0  13210.00 2014-01-01           1  
2     32670.00   21780.0  10890.00 2014-01-06           2  
3     13320.00    8880.0   4440.00 2014-01-06           2  
4     37050.00   24700.0  12350.00 2014-01-06           2  
..         ...       ...       ...        ...         ...  
695  631125.00  618750.0  12375.00 2014-01-03           5  
696  139230.00  136500.0   2730.00 2014-01-10           5  
697    8139.60    6840.0   1299.60 2014-01-02           1  
698    4301.85    3615.0    686.85 2014-01-04           1  
699   18421.20    5418.0  13003.20 2014-01-05           3  

[700 rows x 14 columns]

- Pada tabel diatas diketahui bahwa kolom id segment merupakan foreign key dari tabel segment

- DataFrame yang digabungkan berisi baris dari DataFrame (df) dan kolom yang cocok dari DataFrame kanan(df1)

In [38]:
# Mengganti nama kolom segment menjadi segment id
merge_df['segment']=merge_df['id segment']

In [39]:
# Menghapus kolom id segment 
merge_df.drop(columns=['id segment'], inplace=True)

In [40]:
 # Menampilkan hasil gabungan DataFrame Segment
merge_df

segment                   country    product discount band  units sold  \
0          1                    Canada  Carretera          None      1618.5   
1          1                   Germany  Carretera          None      1321.0   
2          2                    France  Carretera          None      2178.0   
3          2                   Germany  Carretera          None       888.0   
4          2                    Mexico  Carretera          None      2470.0   
..       ...                       ...        ...           ...         ...   
695        5                    France   Amarilla          High      2475.0   
696        5                    Mexico   Amarilla          High       546.0   
697        1                    Mexico    Montana          High      1368.0   
698        1                    Canada      Paseo          High       723.0   
699        3  United States of America        VTT          High      1806.0   

     manufacturing price  sale price  gross sales  discounts      sales  \
0                    3.0        20.0      32370.0       0.00   32370.00   
1                    3.0        20.0      26420.0       0.00   26420.00   
2                    3.0        15.0      32670.0       0.00   32670.00   
3                    3.0        15.0      13320.0       0.00   13320.00   
4                    3.0        15.0      37050.0       0.00   37050.00   
..                   ...         ...          ...        ...        ...   
695                260.0       300.0     742500.0  111375.00  631125.00   
696                260.0       300.0     163800.0   24570.00  139230.00   
697                  5.0         7.0       9576.0    1436.40    8139.60   
698                 10.0         7.0       5061.0     759.15    4301.85   
699                250.0        12.0      21672.0    3250.80   18421.20   

         cogs    profit       date  
0     16185.0  16185.00 2014-01-01  
1     13210.0  13210.00 2014-01-01  
2     21780.0  10890.00 2014-01-06  
3      8880.0   4440.00 2014-01-06  
4     24700.0  12350.00 2014-01-06  
..        ...       ...        ...  
695  618750.0  12375.00 2014-01-03  
696  136500.0   2730.00 2014-01-10  
697    6840.0   1299.60 2014-01-02  
698    3615.0    686.85 2014-01-04  
699    5418.0  13003.20 2014-01-05  

[700 rows x 13 columns]

In [41]:
'''
Menggabungkan dua DataFrame berdasarkan kolom 'country' menggunakan left join
'''

merge_df2 = pd.merge(merge_df,df2, on ='country', how='left')
merge_df2 

segment                   country    product discount band  units sold  \
0          1                    Canada  Carretera          None      1618.5   
1          1                   Germany  Carretera          None      1321.0   
2          2                    France  Carretera          None      2178.0   
3          2                   Germany  Carretera          None       888.0   
4          2                    Mexico  Carretera          None      2470.0   
..       ...                       ...        ...           ...         ...   
695        5                    France   Amarilla          High      2475.0   
696        5                    Mexico   Amarilla          High       546.0   
697        1                    Mexico    Montana          High      1368.0   
698        1                    Canada      Paseo          High       723.0   
699        3  United States of America        VTT          High      1806.0   

     manufacturing price  sale price  gross sales  discounts      sales  \
0                    3.0        20.0      32370.0       0.00   32370.00   
1                    3.0        20.0      26420.0       0.00   26420.00   
2                    3.0        15.0      32670.0       0.00   32670.00   
3                    3.0        15.0      13320.0       0.00   13320.00   
4                    3.0        15.0      37050.0       0.00   37050.00   
..                   ...         ...          ...        ...        ...   
695                260.0       300.0     742500.0  111375.00  631125.00   
696                260.0       300.0     163800.0   24570.00  139230.00   
697                  5.0         7.0       9576.0    1436.40    8139.60   
698                 10.0         7.0       5061.0     759.15    4301.85   
699                250.0        12.0      21672.0    3250.80   18421.20   

         cogs    profit       date  id country  
0     16185.0  16185.00 2014-01-01           1  
1     13210.0  13210.00 2014-01-01           2  
2     21780.0  10890.00 2014-01-06           3  
3      8880.0   4440.00 2014-01-06           2  
4     24700.0  12350.00 2014-01-06           4  
..        ...       ...        ...         ...  
695  618750.0  12375.00 2014-01-03           3  
696  136500.0   2730.00 2014-01-10           4  
697    6840.0   1299.60 2014-01-02           4  
698    3615.0    686.85 2014-01-04           1  
699    5418.0  13003.20 2014-01-05           5  

[700 rows x 14 columns]

- Pada tabel diatas diketahui bahwa kolom id country merupakan foreign key dari tabel country

- DataFrame yang digabungkan berisi baris dari DataFrame (merge_df) dan kolom yang cocok dari DataFrame kanan(df2)

In [42]:
# Replace nilai kolom id pada kolom country
merge_df2['country']=merge_df2['id country']

# Menghapus kolom id country
merge_df2.drop(columns=['id country'], inplace=True)

In [43]:
 # Menampilkan hasil gabungan DataFrame Country
merge_df2

segment  country    product discount band  units sold  \
0          1        1  Carretera          None      1618.5   
1          1        2  Carretera          None      1321.0   
2          2        3  Carretera          None      2178.0   
3          2        2  Carretera          None       888.0   
4          2        4  Carretera          None      2470.0   
..       ...      ...        ...           ...         ...   
695        5        3   Amarilla          High      2475.0   
696        5        4   Amarilla          High       546.0   
697        1        4    Montana          High      1368.0   
698        1        1      Paseo          High       723.0   
699        3        5        VTT          High      1806.0   

     manufacturing price  sale price  gross sales  discounts      sales  \
0                    3.0        20.0      32370.0       0.00   32370.00   
1                    3.0        20.0      26420.0       0.00   26420.00   
2                    3.0        15.0      32670.0       0.00   32670.00   
3                    3.0        15.0      13320.0       0.00   13320.00   
4                    3.0        15.0      37050.0       0.00   37050.00   
..                   ...         ...          ...        ...        ...   
695                260.0       300.0     742500.0  111375.00  631125.00   
696                260.0       300.0     163800.0   24570.00  139230.00   
697                  5.0         7.0       9576.0    1436.40    8139.60   
698                 10.0         7.0       5061.0     759.15    4301.85   
699                250.0        12.0      21672.0    3250.80   18421.20   

         cogs    profit       date  
0     16185.0  16185.00 2014-01-01  
1     13210.0  13210.00 2014-01-01  
2     21780.0  10890.00 2014-01-06  
3      8880.0   4440.00 2014-01-06  
4     24700.0  12350.00 2014-01-06  
..        ...       ...        ...  
695  618750.0  12375.00 2014-01-03  
696  136500.0   2730.00 2014-01-10  
697    6840.0   1299.60 2014-01-02  
698    3615.0    686.85 2014-01-04  
699    5418.0  13003.20 2014-01-05  

[700 rows x 13 columns]

In [44]:
'''
Menggabungkan dua DataFrame berdasarkan kolom 'product' menggunakan left join
'''

merge_df3 = pd.merge(merge_df2,df3, on ='product', how='left')
merge_df3

segment  country    product discount band  units sold  \
0          1        1  Carretera          None      1618.5   
1          1        2  Carretera          None      1321.0   
2          2        3  Carretera          None      2178.0   
3          2        2  Carretera          None       888.0   
4          2        4  Carretera          None      2470.0   
..       ...      ...        ...           ...         ...   
695        5        3   Amarilla          High      2475.0   
696        5        4   Amarilla          High       546.0   
697        1        4    Montana          High      1368.0   
698        1        1      Paseo          High       723.0   
699        3        5        VTT          High      1806.0   

     manufacturing price  sale price  gross sales  discounts      sales  \
0                    3.0        20.0      32370.0       0.00   32370.00   
1                    3.0        20.0      26420.0       0.00   26420.00   
2                    3.0        15.0      32670.0       0.00   32670.00   
3                    3.0        15.0      13320.0       0.00   13320.00   
4                    3.0        15.0      37050.0       0.00   37050.00   
..                   ...         ...          ...        ...        ...   
695                260.0       300.0     742500.0  111375.00  631125.00   
696                260.0       300.0     163800.0   24570.00  139230.00   
697                  5.0         7.0       9576.0    1436.40    8139.60   
698                 10.0         7.0       5061.0     759.15    4301.85   
699                250.0        12.0      21672.0    3250.80   18421.20   

         cogs    profit       date  id product  
0     16185.0  16185.00 2014-01-01           1  
1     13210.0  13210.00 2014-01-01           1  
2     21780.0  10890.00 2014-01-06           1  
3      8880.0   4440.00 2014-01-06           1  
4     24700.0  12350.00 2014-01-06           1  
..        ...       ...        ...         ...  
695  618750.0  12375.00 2014-01-03           6  
696  136500.0   2730.00 2014-01-10           6  
697    6840.0   1299.60 2014-01-02           2  
698    3615.0    686.85 2014-01-04           3  
699    5418.0  13003.20 2014-01-05           5  

[700 rows x 14 columns]

- Pada tabel diatas diketahui bahwa kolom id product merupakan foreign key dari tabel product

- DataFrame yang digabungkan berisi baris dari DataFrame (merge_df2) dan kolom yang cocok dari DataFrame kanan(df3) 

In [45]:
# Replace nilai kolom id pada kolom product
merge_df3['product']=merge_df3['id product']

# Menghapus kolom id produdct
merge_df3.drop(columns=['id product'], inplace=True)

In [46]:
 # Menampilkan hasil gabungan DataFrame Product
merge_df3.head()

segment  country  product discount band  units sold  manufacturing price  \
0        1        1        1          None      1618.5                  3.0   
1        1        2        1          None      1321.0                  3.0   
2        2        3        1          None      2178.0                  3.0   
3        2        2        1          None       888.0                  3.0   
4        2        4        1          None      2470.0                  3.0   

   sale price  gross sales  discounts    sales     cogs   profit       date  
0        20.0      32370.0        0.0  32370.0  16185.0  16185.0 2014-01-01  
1        20.0      26420.0        0.0  26420.0  13210.0  13210.0 2014-01-01  
2        15.0      32670.0        0.0  32670.0  21780.0  10890.0 2014-01-06  
3        15.0      13320.0        0.0  13320.0   8880.0   4440.0 2014-01-06  
4        15.0      37050.0        0.0  37050.0  24700.0  12350.0 2014-01-06

In [47]:
'''
Menggabungkan dua DataFrame berdasarkan kolom 'discount band' menggunakan left join
'''
merge_df4= merge_df3.merge(df4, on ='discount band', how ='left')
merge_df4

segment  country  product discount band  units sold  manufacturing price  \
0          1        1        1          None      1618.5                  3.0   
1          1        2        1          None      1321.0                  3.0   
2          2        3        1          None      2178.0                  3.0   
3          2        2        1          None       888.0                  3.0   
4          2        4        1          None      2470.0                  3.0   
..       ...      ...      ...           ...         ...                  ...   
695        5        3        6          High      2475.0                260.0   
696        5        4        6          High       546.0                260.0   
697        1        4        2          High      1368.0                  5.0   
698        1        1        3          High       723.0                 10.0   
699        3        5        5          High      1806.0                250.0   

     sale price  gross sales  discounts      sales      cogs    profit  \
0          20.0      32370.0       0.00   32370.00   16185.0  16185.00   
1          20.0      26420.0       0.00   26420.00   13210.0  13210.00   
2          15.0      32670.0       0.00   32670.00   21780.0  10890.00   
3          15.0      13320.0       0.00   13320.00    8880.0   4440.00   
4          15.0      37050.0       0.00   37050.00   24700.0  12350.00   
..          ...          ...        ...        ...       ...       ...   
695       300.0     742500.0  111375.00  631125.00  618750.0  12375.00   
696       300.0     163800.0   24570.00  139230.00  136500.0   2730.00   
697         7.0       9576.0    1436.40    8139.60    6840.0   1299.60   
698         7.0       5061.0     759.15    4301.85    3615.0    686.85   
699        12.0      21672.0    3250.80   18421.20    5418.0  13003.20   

          date  id discount band  
0   2014-01-01                 1  
1   2014-01-01                 1  
2   2014-01-06                 1  
3   2014-01-06                 1  
4   2014-01-06                 1  
..         ...               ...  
695 2014-01-03                 4  
696 2014-01-10                 4  
697 2014-01-02                 4  
698 2014-01-04                 4  
699 2014-01-05                 4  

[700 rows x 14 columns]

- Pada tabel diatas diketahui bahwa kolom id discount band merupakan foreign key dari tabel discount band

- DataFrame yang digabungkan berisi baris dari DataFrame (merge_df3) dan kolom yang cocok dari DataFrame kanan(df4) 

In [48]:
# Replace nilai kolom id pada kolom discount band
merge_df4['discount band']=merge_df4['id discount band']

# Menghapus kolom id produdct
merge_df4.drop(columns=['id discount band'], inplace=True)

In [49]:
 # Menampilkan hasil gabungan DataFrame Discount Band
merge_df4

segment  country  product  discount band  units sold  \
0          1        1        1              1      1618.5   
1          1        2        1              1      1321.0   
2          2        3        1              1      2178.0   
3          2        2        1              1       888.0   
4          2        4        1              1      2470.0   
..       ...      ...      ...            ...         ...   
695        5        3        6              4      2475.0   
696        5        4        6              4       546.0   
697        1        4        2              4      1368.0   
698        1        1        3              4       723.0   
699        3        5        5              4      1806.0   

     manufacturing price  sale price  gross sales  discounts      sales  \
0                    3.0        20.0      32370.0       0.00   32370.00   
1                    3.0        20.0      26420.0       0.00   26420.00   
2                    3.0        15.0      32670.0       0.00   32670.00   
3                    3.0        15.0      13320.0       0.00   13320.00   
4                    3.0        15.0      37050.0       0.00   37050.00   
..                   ...         ...          ...        ...        ...   
695                260.0       300.0     742500.0  111375.00  631125.00   
696                260.0       300.0     163800.0   24570.00  139230.00   
697                  5.0         7.0       9576.0    1436.40    8139.60   
698                 10.0         7.0       5061.0     759.15    4301.85   
699                250.0        12.0      21672.0    3250.80   18421.20   

         cogs    profit       date  
0     16185.0  16185.00 2014-01-01  
1     13210.0  13210.00 2014-01-01  
2     21780.0  10890.00 2014-01-06  
3      8880.0   4440.00 2014-01-06  
4     24700.0  12350.00 2014-01-06  
..        ...       ...        ...  
695  618750.0  12375.00 2014-01-03  
696  136500.0   2730.00 2014-01-10  
697    6840.0   1299.60 2014-01-02  
698    3615.0    686.85 2014-01-04  
699    5418.0  13003.20 2014-01-05  

[700 rows x 13 columns]

In [50]:
# Mengganti nama kolom
merge_df4.rename(columns={'segment': 'segment id', 'country': 'country id', 'product': 'product id', 'discount band': 'discount band id'}, inplace=True)
merge_df4

segment id  country id  product id  discount band id  units sold  \
0             1           1           1                 1      1618.5   
1             1           2           1                 1      1321.0   
2             2           3           1                 1      2178.0   
3             2           2           1                 1       888.0   
4             2           4           1                 1      2470.0   
..          ...         ...         ...               ...         ...   
695           5           3           6                 4      2475.0   
696           5           4           6                 4       546.0   
697           1           4           2                 4      1368.0   
698           1           1           3                 4       723.0   
699           3           5           5                 4      1806.0   

     manufacturing price  sale price  gross sales  discounts      sales  \
0                    3.0        20.0      32370.0       0.00   32370.00   
1                    3.0        20.0      26420.0       0.00   26420.00   
2                    3.0        15.0      32670.0       0.00   32670.00   
3                    3.0        15.0      13320.0       0.00   13320.00   
4                    3.0        15.0      37050.0       0.00   37050.00   
..                   ...         ...          ...        ...        ...   
695                260.0       300.0     742500.0  111375.00  631125.00   
696                260.0       300.0     163800.0   24570.00  139230.00   
697                  5.0         7.0       9576.0    1436.40    8139.60   
698                 10.0         7.0       5061.0     759.15    4301.85   
699                250.0        12.0      21672.0    3250.80   18421.20   

         cogs    profit       date  
0     16185.0  16185.00 2014-01-01  
1     13210.0  13210.00 2014-01-01  
2     21780.0  10890.00 2014-01-06  
3      8880.0   4440.00 2014-01-06  
4     24700.0  12350.00 2014-01-06  
..        ...       ...        ...  
695  618750.0  12375.00 2014-01-03  
696  136500.0   2730.00 2014-01-10  
697    6840.0   1299.60 2014-01-02  
698    3615.0    686.85 2014-01-04  
699    5418.0  13003.20 2014-01-05  

[700 rows x 13 columns]

In [51]:
# Menampilkan ringkasan informasi setelah penggabungan DataFrame
merge_df4.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 700 entries, 0 to 699
Data columns (total 13 columns):
 #   Column               Non-Null Count  Dtype         
---  ------               --------------  -----         
 0   segment id           700 non-null    int64         
 1   country id           700 non-null    int64         
 2   product id           700 non-null    int64         
 3   discount band id     700 non-null    int64         
 4   units sold           700 non-null    float64       
 5   manufacturing price  700 non-null    float64       
 6   sale price           700 non-null    float64       
 7   gross sales          700 non-null    float64       
 8   discounts            700 non-null    float64       
 9   sales                700 non-null    float64       
 10  cogs                 700 non-null    float64       
 11  profit               700 non-null    float64       
 12  date                 700 non-null    datetime64[ns]
dtypes: datetime64[ns](1), float64(8), i

Berdasarkan hasil tersebut diketahui bahwa terdapat 700 baris dan 13 kolom setelah dilakukan data cleaning dan data normalization. Tipe data setelah penggabungan DataFrame juga berubah sesuai dengan hasil yang tertera.

In [54]:
# menghapus index di kolom segment
df1.set_index('id segment', inplace=True)

# simpan tabel country
df1.to_csv('tabelsegmentnew.csv', index=False)

Menghapus index DataFrame Segment untuk di save dalam file csv, karena saat di olah di postgres akanmenerapkan auto incremental untuk pengisian primay key.

In [55]:
df1

segment
id segment                  
1                 Government
2                  Midmarket
3           Channel Partners
4                 Enterprise
5             Small Business

In [56]:
# menghapus index di kolom country
df2.set_index('id country', inplace=True)

# simpan tabel country
df2.to_csv('tabelcountrynew.csv', index=False)

Menghapus index DataFrame Country untuk di save dalam file csv, karena saat di olah di postgres akanmenerapkan auto incremental untuk pengisian primay key.

In [57]:
df2

country
id country                          
1                             Canada
2                            Germany
3                             France
4                             Mexico
5           United States of America

In [58]:
# menghapus index di kolom product
df3.set_index('id product', inplace=True)

# simpan tabel country
df3.to_csv('tabelproductnew.csv', index=False)

Menghapus index DataFrame Product untuk di save dalam file csv, karena saat di olah di postgres akanmenerapkan auto incremental untuk pengisian primay key.

In [59]:
df3

product
id product           
1           Carretera
2             Montana
3               Paseo
4                Velo
5                 VTT
6            Amarilla

In [60]:
# menghapus index di kolom discount band
df4.set_index('id discount band', inplace=True)

# simpan tabel country
df4.to_csv('tabeldiscountbandnew.csv', index=False)

Menghapus index DataFrame Discount Band untuk di save dalam file csv, karena saat di olah di postgres akanmenerapkan auto incremental untuk pengisian primay key.

In [61]:
df4

discount band
id discount band              
1                         None
2                          Low
3                       Medium
4                         High

In [ ]:
# Menampilkan kembali data yang sudah di gabungkan
merge_df4

segment id  country id  product id  discount band id  units sold  \
0             1           1           1                 1      1618.5   
1             1           2           1                 1      1321.0   
2             2           3           1                 1      2178.0   
3             2           2           1                 1       888.0   
4             2           4           1                 1      2470.0   
..          ...         ...         ...               ...         ...   
695           5           3           6                 4      2475.0   
696           5           4           6                 4       546.0   
697           1           4           2                 4      1368.0   
698           1           1           3                 4       723.0   
699           3           5           5                 4      1806.0   

     manufacturing price  sale price  gross sales  discounts      sales  \
0                    3.0        20.0      32370.0       0.00   32370.00   
1                    3.0        20.0      26420.0       0.00   26420.00   
2                    3.0        15.0      32670.0       0.00   32670.00   
3                    3.0        15.0      13320.0       0.00   13320.00   
4                    3.0        15.0      37050.0       0.00   37050.00   
..                   ...         ...          ...        ...        ...   
695                260.0       300.0     742500.0  111375.00  631125.00   
696                260.0       300.0     163800.0   24570.00  139230.00   
697                  5.0         7.0       9576.0    1436.40    8139.60   
698                 10.0         7.0       5061.0     759.15    4301.85   
699                250.0        12.0      21672.0    3250.80   18421.20   

         cogs    profit       date  
0     16185.0  16185.00 2014-01-01  
1     13210.0  13210.00 2014-01-01  
2     21780.0  10890.00 2014-01-06  
3      8880.0   4440.00 2014-01-06  
4     24700.0  12350.00 2014-01-06  
..        ...       ...        ...  
695  618750.0  12375.00 2014-01-03  
696  136500.0   2730.00 2014-01-10  
697    6840.0   1299.60 2014-01-02  
698    3615.0    686.85 2014-01-04  
699    5418.0  13003.20 2014-01-05  

[700 rows x 13 columns]

In [62]:
# simpan tabel utama
merge_df4.to_csv('tabelfinance.csv', index=False)

# **6. Relational Database & SQL**

##### a. Menampilkan tabel mengenai total profit tiap jenis segmentasi dengan kondisi data yang tidak diskon




SELECT

s."segment",  						-- Menampilkan data dari tabel segment

SUM(f."profit") AS "total profit"  	-- Menghitung total profit dari tabel finance

FROM finance AS f 					-- Diambil dari data tabel finance

JOIN segment AS s 					-- Di gabungkan kan dengan tabel segment

ON s.id = f."segment id" 			 -- Menghubungkan kolom id tabel segment dengan kolom segment id tabel finance

WHERE "discounts" !=0 				-- Dengan kondisi tidak mengambil data yang diskon

GROUP BY s."segment"				-- Dikelompokan berdasarkan tabel segment
	   
ORDER BY SUM(f."profit") DESC;		-- Di urutukan berdasarkan total profit terbesar

In [ ]:
# Menampilkan jawaban a dari query SQL yang sudah dibuat
Jawaban_5a = pd.read_csv ('Jawaban 5a.csv')
display(Jawaban_5a)

segment  total profit
0        Government   10270950.18
1    Small Business    3848268.50
2  Channel Partners    1150321.14
3         Midmarket     563253.09
4        Enterprise    -675545.62

- Dari hasil analisis diketahui bahwa tiap segmentasi yang memiliki total profit terbanyak (diluar data diskon) adalah segementasi "Goverment" sebanyak 10270950.18. Posisi kedua ada segmentasi "Small Business" sebanyak 3848268.50 dan posisi kediga ada segmentasi "Channel Partners" sebanyak 1150321.14.

- Adapun segmentasi yang memiliki total profit minus atau mengalami kerugia ada di segmentasi "Enterprise" sebanyak -675545.62.

- Dari data tersebut dapat dilakukan evaluasi tiap segmentasi untuk meningkatkan total profit tiap segmentasi.

##### b. Menampilkan tabel mengenai nilai rata-rata, min, dan max dari sales masing-masing negara



SELECT

c."country", 				          -- Menampilkan data dari tabel country

AVG(f."sales") AS "rata-rata sales", -- Menghitung nilai rata-rata dari sales di tabel finance

MIN(f."sales") AS "minimum sales",	 -- Menghitung nilai minimum dari sales di tabel finance

MAX(f."sales") AS "maksimal sales"    -- Menghitung nilai maksimal dari sales di tabel finance

FROM finance AS f 			          -- Diambil dari data tabel finance	

JOIN country AS c						  -- Di gabungkan kan dengan tabel country

ON c.id = f."country id"				  -- Menghubungkan kolom id tabel country dengan kolom country id tabel finance

GROUP BY c."country"					  -- Dikelompokan berdasarkan tabel country

ORDER BY 2 DESC;			  			  -- Di urutkan berdasarkan rata-rata sales terbesar


In [ ]:
# Menampilkan jawaban b dari query SQL yang sudah dibuat
Jawaban_5b = pd.read_csv ('Jawaban 5b.csv')
display(Jawaban_5b)

country  rata-rata sales  minimum sales  maksimal sales
0  United States of America    178784.501286        1730.54       1159200.0
1                    Canada    177768.963500        1655.08       1038082.5
2                    France    173958.373500        1763.86        962500.0
3                   Germany    167895.291571        1685.60       1017338.0
4                    Mexico    149638.229357        2508.66        848172.5

- Dari hasil analisi menunjukkan bahwa rata-rata sales atau penjualan terbanyak berdasarkan negara ada di negara "United States of America" sebanyak 178784.501286 dan rata-rata sales terendah ada di negara "Mexico" sebanyak 149638.229357.

- United States of America memiliki nilai maksimal sales paling tinggi sebesar 1159200.0.

- Canada memiliki nila minimum sales paling rendah sebesar 1655.08. Jika target tidak memenuhi nilai mimnimum sales yang dicapai, maka perusahaan akan mengalami kerugian.

- Dari tabel tersebut dapat diketahui nilai sales/penjualan terbanyak dan terendah tiap masing-masing negara, hasil tabel ini dapat dijadikan bahan evaluasi untuk meningkatkan nilai penjualan tiap negara. 

### KESIMPULAN

Sebagai seorang Data Engineer untuk mengelola data diperlukan pembuatan Database. Proses pembuatan Database memerlukan beberapa proses untuk mempersiapkan data tersebut bersih dan siap untuk dimasukkan ke dalam database. Adapun langkah-langkah yang harus dilakukan, sebagai berikut:

1. Melakukan Data Explorartion
    Pada bagian ini diawali dengan load file dari sumber file yang diberikan. Lalu melakukan pencarian data untuk mengetahui ringkasan informasi dari data tersebut, seperti berapa banyak baris dan kolom. Selain itu, untuk mengetahui apakah data tersebut ada missing value atau whitespace sehingga data tersebut harus dilakukan Data Cleaning.

2. Melakukan Data Cleaning
    Setelah mengetahui hal-hal yang harus di hapus untuk menghindari error saat analisis data dan dilakukan handling missing value. Proses ini sangat diperlukan untuk mempersiapkan data yang siap dimasukkan ke dalam database. Seperti menyesuaikan tipe data nya.

3. Melakukan Data Normalization
    Tahap ini melakukan olah data sampai dengan 3NF dengan memperhatikan penggunaan primary key dan foreign key antar tabel

4. Membuat Database
    Setelah data sudah bersih dan siap dimasukkan ke dalam database, maka dilakukan pembuatan database dan tabel beserta menginput data dengan PostgresSQL server yang dilakukan dengan perintah TCL (Transaction control language). Setelah itu dilakukan pembuatan user untuk mengakses beberapa command query yang diberikan oleh admin.

5. Database siap dikelola untuk mendapatkan insight atau hasil temuan dari analisis data yang sudah dikumpulan. Insight tersebut digunakan untuk memprediksi dan mengembangkan perusahaan untuk yang lebih baik. 